File: src/0015.3Sum-Array/3sum.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int> > threeSum(vector<int>& nums) {
        if(nums.size()<3) return {};
        vector<vector<int> > result;
        sort(nums.begin(), nums.end());
        for(int i = 0; i < (nums.size()-2); ++i)
        {
            if(nums[i] > 0) break;
            if(i>0 && nums[i] == nums[i-1]) continue;
            int ll = i+1;
            int rr = nums.size()-1;
            while(ll < rr)
            {
                int sum = nums[i] + nums[ll] + nums[rr];
                if(sum == 0)
                {
                    result.push_back({nums[i],nums[ll],nums[rr]});
                    ll++;
                    rr--;
                    while(ll < rr && nums[ll] == nums[ll-1]) ll++;
                    while(ll < rr && nums[rr] == nums[rr+1]) rr--;
                }
                else if(sum > 0)
                {
                    rr--;
                }
                else if(sum < 0)
                {
                    ll++;
                }
            }
        }
        return result;
    }
};

int main()
{
    vector<int> nums = {1,2,3,0,-1};
    Solution A;
    vector<vector<int> > ans = A.threeSum(nums);
    for(int i = 0; i<ans.size(); ++i)
    {
        for(int j = 0; j<ans[i].size(); ++j)
        {
            cout << ans[i][j] << endl;
        }
    }
    
    return 0;
}

File: src/0015.3Sum-Array/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        for (int k = 0; k < nums.length - 2; k++) {
            if (nums[k] > 0) {
                return res;
            }
            if (k > 0 && nums[k] == nums[k-1]) {
                continue;
            }
            int i = k + 1;
            int j = nums.length - 1;
            while (i < j) {
                int tmp = nums[k] + nums[i] + nums[j];
                if (tmp == 0) {
                    res.add(Arrays.asList(nums[k],nums[i],nums[j]));
                    while (i < j && nums[i] == nums[i+1]) {
                        i++;
                    }
                    while (i < j && nums[j] == nums[j-1]) {
                        j--;
                    }
                    i++;
                    j--;
                } else if(tmp > 0) {
                    j--;
                } else {
                    i++;
                }
            }
        }
        return res;
    }
}